src/graphnet/models/gnn/dynedge_kaggle_tito.py

```diff
@@ -0,0 +1,253 @@
+        )
+        nb_latent_features = last_posting_layer_output_dim * nb_poolings
+        nb_latent_features += self._nb_global_variables
+
+        readout_layers = []
+        layer_sizes = [nb_latent_features] + list(self._readout_layer_sizes)
+        for nb_in, nb_out in zip(layer_sizes[:-1], layer_sizes[1:]):
+            readout_layers.append(torch.nn.Linear(nb_in, nb_out))
+            readout_layers.append(self._activation)
+
+        self._readout = torch.nn.Sequential(*readout_layers)
+
+        # Transformer layer(s)
+        encoder_layer = TransformerEncoderLayer(
+            d_model=last_posting_layer_output_dim,
+            nhead=8,
+            batch_first=True,
+            norm_first=False,
+        )
+        self._transformer_encoder = TransformerEncoder(
+            encoder_layer, num_layers=0
+        )
+
+    def _global_pooling(self, x: Tensor, batch: LongTensor) -> Tensor:
+        """Perform global pooling."""
+        assert self._global_pooling_schemes
+        pooled = []
+        for pooling_scheme in self._global_pooling_schemes:
+            pooling_fn = GLOBAL_POOLINGS[pooling_scheme]
+            pooled_x = pooling_fn(x, index=batch, dim=0)
+            if isinstance(pooled_x, tuple) and len(pooled_x) == 2:
+                # `scatter_{min,max}`, which return also an argument, vs.
+                # `scatter_{mean,sum}`
+                pooled_x, _ = pooled_x
+            pooled.append(pooled_x)
+
+        return torch.cat(pooled, dim=1)
+
+    def _calculate_global_variables(
+        self,
+        x: Tensor,
+        edge_index: LongTensor,
+        batch: LongTensor,
+        *additional_attributes: Tensor,
+    ) -> Tensor:
+        """Calculate global variables."""
+        # Calculate homophily (scalar variables)
+        h_x, h_y, h_z, h_t = calculate_xyzt_homophily(x, edge_index, batch)
+
+        # Calculate mean features
+        global_means = scatter_mean(x, batch, dim=0)
+
+        # Add global variables
+        global_variables = torch.cat(
+            [
+                global_means,
+                h_x,
+                h_y,
+                h_z,
+                h_t,
+            ]
+            + [attr.unsqueeze(dim=1) for attr in additional_attributes],
+            dim=1,
+        )
+
+        return global_variables
+
+    def forward(self, data: Data) -> Tensor:
+        """Apply learnable forward pass."""
+        # Convenience variables
+        x, edge_index, batch = data.x, data.edge_index, data.batch
+
+        global_variables = self._calculate_global_variables(
+            x,
+            edge_index,
+            batch,
+            torch.log10(data.n_pulses),
+        )
+
+        # DynEdge-convolutions
+        for conv_layer in self._conv_layers:
+            x, _edge_index = conv_layer(x, data.edge_index, batch)
+
+        # Transformer convolutions
+        x, mask = to_dense_batch(x, batch)
+        x = self._transformer_encoder(x, src_key_padding_mask=mask)
```

Comment:
See comment above; `self._transformer_encoder` is set to an empty module list. Are you sure this is needed?